### ISSUE TITLE
Sorting tracks alphabetically is underspecified

### ISSUE THREAD

**Issue by svgeesus** [WebAudio/web-audio-api#1772]
See https://github.com/w3c/i18n-activity/issues/598

**Comment by padenot** [comment]
Let's order by code point. It's not that important, it just needs to be a stable sort.

**Comment by rtoy** [comment]
I am not a Unicode expert but I think sorting by codepoint is wrong. The
Unicode spec defines a default collation algorithm for sorting (
http://unicode.org/reports/tr10/) that would seem a better choice.

But it seems JS sorts strings by codepoint as well, so maybe that's best.

The problem with codepoints is that the mapping between visual glypths and
codepoints is not unique. I guess that's not out problem to solve here.

**Comment by svgeesus** [comment]
I agree it just needs to be

a) stable
b) defined

code point is easy, fast, and meets the needs.

**Comment by svgeesus** [comment]
Note that there was [further discussion of this issue by I18n WG](https://github.com/w3c/i18n-activity/issues/598) on their github. Note that they suggest not doing casefolding (not even ASCII-only casefolding) which then makes this super simple.

**Comment by aphillips** [comment]
As I [mentioned ](https://github.com/w3c/i18n-activity/issues/598#issuecomment-561925317) in our issue, I don't think the text in the the current editor's draft fixes the problem. The term "lexicographic" means roughly the same thing as "alphabetical", whereas I think the intention is to do code-point-by-code-point comparison. In addition, the new text uses the term "code unit", which is probably incorrect.

At issue in the above discussion is whether the sort order is ever visible to human beings. If your goal is to produce a list that looks approximately correct to end users and which uses natural language strings, then you might want to introduce the bells and whistles of Unicode collation. However, my understanding of the issue from @stpeter's comments is that these are mainly opaque IDs, such as UUIDs, and while the ordering needs to be deterministic, it is arbitrary. The cheapest and easiest way to implement this is to do code point comparison. This is also consistent with [HTML string comparison](https://html.spec.whatwg.org/#case-sensitivity-and-string-comparison). As such, I'd rephrase to say:

> Sort the elements in *tracks* based on their id attribute by comparing the values code point by code point.

This seems like too minor a feature to warrant additional health warnings (such as mentioning that the comparison is case sensitive and shouldn't be expected to produce a lexicographical order).

**Comment by rtoy** [comment]
Maybe just remove the text about "lexicographic".  We already say we sort by code units (not code points).

**Comment by rtoy** [comment]
Hmm.  HTML string comparison is by code points.  Javascript strings are in UTF-16, so code point comparison and code unit comparison can be different.

**Comment by padenot** [comment]
We should put https://github.com/w3c/i18n-activity/issues/598#issuecomment-561925317 in the spec. I think the current prose is _not_ what implementations are doing, and that's a mistake in the spec.

**Comment by rtoy** [comment]
Chrome sorts by code units because that's how Javascript [sorts strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort).

But I agree with everything else in that comment.

**Comment by rtoy** [comment]
Teleconf:  Remove "lexicographic".  But still require sorting by code units because that's how Javascript sorts strings.  (Code points and code units are different.)

**Comment by svgeesus** [comment]
@aphillips yes, these are opaque IDs for individual tracks, and are not presented to the user. The intent is simply that implementations all have a consistent track ordering. The term `code unit` is not an error and is due to the fact that Javascript stores strings in UTF-16 and indexes by code unit not code point, so non-BMP characters take two units.

**Comment by aphillips** [comment]
The choice of sorting on code points or code units depends on what one wants to do with the resulting list or array of items. Because this is a list of opaque IDs and the goal is a fast deterministic sort, it makes all kinds of sense to do code *unit* sorting. The change looks good to me.

**Comment by aphillips** [comment]
@svgeesus Please don't remove our label. While the title seems kind of "active" (suggesting the issue isn't resolved), it's actually how we track where our issues are--we don't have an "i18n-resolved" label or such to keep the label clutter down. You are more than welcome to close this issue once you're done with it.

**Comment by svgeesus** [comment]
@aphillips aha, sorry! Understood.

**Comment by rtoy** [comment]
When the issue is closed, I don't think it's necessary.  Might be handy if we ever need to look at closed issues to see which were blockers.

**Comment by himorin** [comment]
Just note: W3C i18n WG closed our tracker with actions taken satisfactory. (please not remove i18n-needs-resolution label which is used by our bot and tracking system)
